src/data/fun_like.lean

```diff
@@ -0,0 +1,172 @@
+my_hom_class.map_op
+
+-- You can replace `my_hom.fun_like` with the below instance, or keep both:
+instance : my_hom_class (my_hom A B) A B :=
+{ coe := my_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_op := my_hom.map_op' }
+
+-- [Insert `has_coe_to_fun`, `to_fun_eq_coe`, `ext` and `copy` here]
+```
+
+The second step is to add instances of your new `my_hom_class` for all types extending `my_hom`.
+Typically, you can just declare a new class analogous to `my_hom_class`:
+
+```
+structure cooler_hom (A B : Type*) [cool_class A] [cool_class B]
+  extends my_hom A B :=
+(map_cool' : to_fun cool_class.cool = cool_class.cool)
+
+class cooler_hom_class (F : Type*) (A B : out_param $ Type*) [cool_class A] [cool_class B]
+  extends my_hom_class F A B :=
+(map_cool : ∀ (f : F), f cool_class.cool = cool_class.cool)
+
+@[simp] lemma map_cool {F A B : Type*} [cool_class A] [cool_class B] [cooler_hom_class F A B]
+  (f : F) : f cool_class.cool = cool_class.cool :=
+my_hom_class.map_op
+
+-- You can also replace `my_hom.fun_like` with the below instance:
+instance : cool_hom_class (cool_hom A B) A B :=
+{ coe := cool_hom.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr',
+  map_op := cool_hom.map_op',
+  map_cool := cool_hom.map_cool' }
+
+-- [Insert `has_coe_to_fun`, `to_fun_eq_coe`, `ext` and `copy` here]
+```
+
+Then any declaration taking a specific type of morphisms as parameter can instead take the
+class you just defined:
+```
+-- Compare with: lemma do_something (f : my_hom A B) : sorry := sorry
+lemma do_something {F : Type*} [my_hom_class F A B] (f : F) : sorry := sorry
+```
+
+This means anything set up for `my_hom`s will automatically work for `cool_hom_class`es,
+and defining `cool_hom_class` only takes a constant amount of effort,
+instead of linearly increasing the work per `my_hom`-related declaration.
+
+-/
+
+-- This instance should have low priority, to ensure we follow the chain
+-- `fun_like → to_fun → has_coe_to_fun`
+attribute [instance, priority 10] coe_fn_trans
+
+/-- The class `fun_like F α β` expresses that terms of type `F` have an
+injective coercion to functions from `α` to `β`.
+
+This typeclass is used in the definition of the homomorphism typeclasses,
+such as `zero_hom_class`, `mul_hom_class`, `monoid_hom_class`, ....
+-/
+class fun_like (F : Sort*) (α β : out_param Sort*) :=
+(coe : F → (α → β))
+(coe_injective' : function.injective coe)
+
+variables (F α β : Sort*)
+
+namespace fun_like
+
+variables {F α β} [i : fun_like F α β]
+
+include i
+
+@[priority 100, -- Give this a priority between `coe_fn_trans` and the default priority
+  nolint dangerous_instance] -- `α` and `β` are out_params, so this instance should not be dangerous
+instance : has_coe_to_fun F (λ _, α → β) := { coe := fun_like.coe }
+
+-- Unfortunately, the elaborator is not smart enough that we can write this as
+-- `function.injective (coe_fn : F → α → β)`
```

Comment:
There are two possible tricks that are/were used to state injectivity of `coe_fn`:
1. Write `@function.injective F (α → β) coe_fn`.
2. Write `function.injective (λ (f : F) (a : α), f a)`.